js/adapt-stateful-session.js

```diff
@@ -1,227 +1,259 @@
-        this.listenTo(Adapt, 'questionView:recordInteraction', this.onQuestionRecordInteraction);
+      if (sessionPairs.q && this._shouldStoreResponses) {
+        this._componentSerializer.deserialize(sessionPairs.q);
       }
+    }
 
-      this.listenTo(Adapt.blocks, 'change:_isComplete', this.onBlockComplete);
+    setupEventListeners() {
+      const debouncedSaveSession = _.debounce(this.saveSessionState.bind(this), 1);
+      this.listenTo(Adapt.data, 'change:_isComplete', debouncedSaveSession);
+      if (this._shouldStoreResponses) {
+        this.listenTo(Adapt.data, 'change:_isSubmitted change:_userAnswer', debouncedSaveSession);
+      }
       this.listenTo(Adapt, {
-        'assessment:complete': this.onAssessmentComplete,
         'app:languageChanged': this.onLanguageChanged,
+        'questionView:recordInteraction': this.onQuestionRecordInteraction,
+        'assessment:complete': this.onAssessmentComplete,
         'tracking:complete': this.onTrackingComplete
       });
-      this.listenTo(Adapt.course, 'change:_isComplete', this.saveSessionState);
-    },
+      const config = Adapt.spoor.config;
+      const advancedSettings = config._advancedSettings;
+      const shouldCommitOnVisibilityChange = (!advancedSettings ||
+          advancedSettings._commitOnVisibilityChangeHidden !== false) &&
+          document.addEventListener;
+      if (shouldCommitOnVisibilityChange) {
+        document.addEventListener('visibilitychange', this.onVisibilityChange);
+      }
+      $(window).on('beforeunload unload', this.endSession);
+    }
 
-    removeEventListeners: function () {
-      $(window).off('beforeunload unload', this._onWindowUnload);
-      this.stopListening();
-    },
+    saveSessionState() {
+      const courseState = SCORMSuspendData.serialize([
+        Boolean(Adapt.course.get('_isComplete')),
+        Boolean(Adapt.course.get('_isAssessmentPassed'))
+      ]);
+      const componentStates = (this._shouldStoreResponses === true) ?
+        this._componentSerializer.serialize() :
+        '';
+      const sessionPairs = {
+        'c': courseState,
+        'q': componentStates
+      };
+      this.printCompletionInformation();
+      Adapt.offlineStorage.set(sessionPairs);
+    }
 
-    reattachEventListeners: function() {
+    printCompletionInformation() {
+      const courseComplete = Boolean(Adapt.course.get('_isComplete'));
+      const assessmentPassed = Boolean(Adapt.course.get('_isAssessmentPassed'));
+      const trackingIdModels = Adapt.data.filter(model => model.get('_type') === this._trackingIdType && model.has('_trackingId'));
+      const trackingIds = trackingIdModels.map(model => model.get('_trackingId'));
+      if (!trackingIds.length) {
+        Adapt.log.info(`course._isComplete: ${courseComplete}, course._isAssessmentPassed: ${assessmentPassed}, ${this._trackingIdType} completion: no tracking ids found`);
+        return;
+      }
+      const max = _.max(trackingIds);
```

Comment:
Suggest native method:
```suggestion
      const max = Math.max(...trackingIds);
```